r1cs/gadgets/std/src/instantiated/tweedle/curves.rs

```diff
@@ -13,22 +10,90 @@ use crate::{
 pub type TweedleDeeGadget = AffineGadget<TweedledeeParameters, Fq, FqGadget>;
 pub type TweedleDumGadget = AffineGadget<TweedledumParameters, Fr, FrGadget>;
 
-#[test]
-fn test_dee() {
-    crate::groups::test::group_test_with_incomplete_add::<
-        _, _, TweedleDeeGadget
-    >();
-    crate::groups::test::mul_bits_test::<
-        _, _, TweedleDeeGadget
-    >();
-}
+#[cfg(test)]
+mod test {
+    use r1cs_core::ConstraintSystem;
+    use crate::{
+        prelude::*,
+        test_constraint_system::TestConstraintSystem,
+        instantiated::tweedle::{
+            TweedleDeeGadget,
+            TweedleDumGadget,
+        }
+    };
+    use algebra::{UniformRand, AffineCurve, ProjectiveCurve, fields::tweedle::{
+        Fq, Fr,
+    }, curves::tweedle::{
+        dee::Projective as DeeProjective,
+        dum::Projective as DumProjective,
+    }, PrimeField, BigInteger};
+    use rand::thread_rng;
+
+    #[test]
+    fn test_dee() {
+        crate::groups::test::group_test_with_incomplete_add::<
+            _, _, TweedleDeeGadget
+        >();
+        crate::groups::test::mul_bits_test::<
+            _, _, TweedleDeeGadget
+        >();
+    }
+
+    #[test]
+    fn test_dum() {
+        crate::groups::test::group_test_with_incomplete_add::<
+            _, _, TweedleDumGadget
+        >();
+        crate::groups::test::mul_bits_test::<
+            _, _, TweedleDumGadget
+        >();
+    }
+
+    #[test]
+    /// Tests equality of the result enforced by `endo_mul()` 
+    /// with the corresponding scalar multiple. 
+    fn test_endo_dee()
+    {
+        let mut cs = TestConstraintSystem::<Fq>::new();
+
+        let a_native = DeeProjective::rand(&mut thread_rng()).into_affine();
+        let a = TweedleDeeGadget::alloc(&mut cs.ns(|| "generate_a"), || Ok(a_native.into_projective())).unwrap();
+
+        let scalar = Fq::rand(&mut thread_rng());
```

Comment:
In this test, we should restrict `scalar` to 128 bits, which is the length for which the gadget is proven secure.
The same below for the dum.